### PR TITLE
Parse version > 9 in dotnet.py

### PR DESCRIPTION
**Pull request by am11**
A job run in PR https://github.com/dotnet/runtime/pull/106599 is failing like https://dev.azure.com/dnceng-public/public/_build/results?buildId=813428&view=logs&j=3ca38e2b-df6d-5512-8718-3c103788a0f9&t=0d6eb74e-f21d-5060-4e66-099a24a67e5e

```
[2024/09/19 22:41:48][INFO] $ pushd "/mnt/vss/_work/1/performance"
[2024/09/19 22:41:48][INFO] $ popd
[2024/09/19 22:41:48][ERROR] Unexpected error: <class 'RuntimeError'>
[2024/09/19 22:41:48][ERROR] Traceback (most recent call last):
  File "/mnt/vss/_work/1/performance/scripts/run_performance_job.py", line 1187, in main
    run_performance_job(RunPerformanceJobArgs(**args))
  File "/mnt/vss/_work/1/performance/scripts/run_performance_job.py", line 768, in run_performance_job
    ci_setup.main(ci_setup_arguments)
  File "/mnt/vss/_work/1/performance/scripts/ci_setup.py", line 447, in main
    dotnet_version = dotnet.get_dotnet_version_precise(target_framework_moniker, args.cli) if args.dotnet_versions == [] else args.dotnet_versions[0]
  File "/mnt/vss/_work/1/performance/scripts/performance/tracer.py", line 90, in wrapper
    return func(*args, **kwargs)
  File "/mnt/vss/_work/1/performance/scripts/dotnet.py", line 614, in get_dotnet_version_precise
    sdk = get_dotnet_version_from_path(framework, dotnet_path, sdk_path)
  File "/mnt/vss/_work/1/performance/scripts/performance/tracer.py", line 90, in wrapper
    return func(*args, **kwargs)
  File "/mnt/vss/_work/1/performance/scripts/dotnet.py", line 601, in get_dotnet_version_from_path
    raise RuntimeError(
RuntimeError: Unable to determine the .NET SDK used for net9.0
```

### CODE REVIEW

**Comment by am11** [comment]
cc @ilonatommy

**Comment by am11** [comment]
Ah, so it is cloning the wrong repo :(

+ git clone https://github.com/dotnet/performance.git --depth 1 -b main --single-branch

File: scripts/dotnet.py [hunk 1]
```diff
@@ -594,12 +596,16 @@ def get_dotnet_version_from_path(
         # Attempt 2: Increase the minor version by 1 and retry.
         sdk = next((f for f in sdks if f.startswith(
             "{}.{}".format(version.major, version.minor + 1))), None)
+    if not sdk:
+        if version.major == 9:
+            sdk = next((f for f in sdks if f.startswith("10.0")), None)
```
Comment (am11) [inline scripts/dotnet.py:601]:
```suggestion
```

**Comment by am11** [comment]
Thanks, sorry I was in my _just before calling it a weekend_ meeting with my manager 🤣

**Comment by ilonatommy** [comment]
From some reason we're ignoring net10 that we parsed and we fall back to net9, see
https://github.com/dotnet/performance/blob/f6d0cff8bfb8bc0c3eec23b4b07c837e4c5f8abb/scripts/performance/common.py#L214
that logs
```
[2024/09/20 11:47:56][INFO] Writing pipeline variable PERFLAB_Framework with value net9.0
```
in https://dev.azure.com/dnceng-public/public/_build/results?buildId=813887&view=logs&j=63c48da7-769c-5a69-ecd6-caa9e1c7e293&t=8b4a1f4f-91f7-5091-3359-c47019cbfacd&l=86

**Comment by am11** [comment]
noticed this one as well: https://helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-106599-merge-9f744b5760d349c987/x64.micro.net10.0.Partition0/1/console.a0f73dde.log?helixlogtype=result
```
[2024/09/20 12:00:47][INFO] //    * The required .NET Core SDK version 9.0 for runtime moniker WasmNet90 is not installed.
[2024/09/20 12:00:47][INFO] 
[2024/09/20 12:00:47][INFO] // Benchmark Perf_Type.GetType_FullyQualifiedNames: Job-ANOAPY(PowerPlanMode=00000000-0000-0000-0000-000000000000, Runtime=Wasm, Toolchain=Wasm, InvocationCount=1, IterationCount=1, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=15, RunStrategy=ColdStart, UnrollFactor=1, WarmupCount=0) [input=System.Int32[,]]
[2024/09/20 12:00:47][INFO] //    * The required .NET Core SDK version 9.0 for runtime moniker WasmNet90 is not installed.
[2024/09/20 12:00:47][INFO] 
[2024/09/20 12:00:47][INFO] // Benchmark Perf_UInt16.ToString: Job-LAURTV(OutlierMode=DontRemove, PowerPlanMode=00000000-0000-0000-0000-000000000000, Runtime=Wasm, Toolchain=Wasm, InvocationCount=1, IterationCount=1, IterationTime=250ms, MaxIterationCount=20, MemoryRandomization=True, MinIterationCount=15, RunStrategy=ColdStart, UnrollFactor=1, WarmupCount=0) [value=0]
[2024/09/20 12:00:47][INFO] //    * The required .NET Core SDK version 9.0 for runtime moniker WasmNet90 is not installed.
[2024/09/20 12:00:47][INFO] 
[2024/09/20 12:00:47][INFO] // Benchmark Perf_UInt32.TryParse: Job-LAURTV(OutlierMode=DontRemove, PowerPlanMode=00000000-0000-0000-0000-000000000000, Runtime=Wasm, Toolchain=Wasm, InvocationCount=1, IterationCount=1, IterationTime=250ms, MaxIterationCount=20, MemoryRandomization=True, MinIterationCount=15, RunStrategy=ColdStart, UnrollFactor=1, WarmupCount=0) [value=0]
[2024/09/20 12:00:47][INFO] //    * The required .NET Core SDK version 9.0 for runtime moniker WasmNet90 is not installed.
[2024/09/20 12:00:47][INFO] 
[2024/09/20 12:00:47][INFO] // Benchmark Perf_UInt64.Parse: Job-ANOAPY(PowerPlanMode=00000000-0000-0000-0000-000000000000, Runtime=Wasm, Toolchain=Wasm, InvocationCount=1, IterationCount=1, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=15, RunStrategy=ColdStart, UnrollFactor=1, WarmupCount=0) [value=0]
[2024/09/20 12:00:47][INFO] //    * The required .NET Core SDK version 9.0 for runtime moniker WasmNet90 is not installed.
[2024/09/20 12:00:47][INFO] 
[2024/09/20 12:00:47][INFO] // Benchmark Perf_Uri.CombineAbsoluteRelative: Job-ANOAPY(PowerPlanMode=00000000-0000-0000-0000-000000000000, Runtime=Wasm, Toolchain=Wasm, InvocationCount=1, IterationCount=1, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=15, RunStrategy=ColdStart, UnrollFactor=1, WarmupCount=0) [input=/new/path]
[2024/09/20 12:00:47][INFO] //    * The required .NET Core SDK version 9.0 for runtime moniker WasmNet90 is not installed.
[2024/09/20 12:00:47][INFO] 
[2024/09/20 12:00:47][INFO] // Benchmark Perf_Uri.CtorIdnHostPathAndQuery: Job-ANOAPY(PowerPlanMode=00000000-0000-0000-0000-000000000000, Runtime=Wasm, Toolchain=Wasm, InvocationCount=1, IterationCount=1, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=15, RunStrategy=ColdStart, UnrollFactor=1, WarmupCount=0) [input=http://host/p(...)s?key=ünicode [50]]
[2024/09/20 12:00:47][INFO] //    * The required .NET Core SDK version 9.0 for runtime moniker WasmNet90 is not installed.
[2024/09/20 12:00:47][INFO] 
[2024/09/20 12:00:47][INFO] // Benchmark Perf_Version.Parse4: Job-LAURTV(OutlierMode=DontRemove, PowerPlanMode=00000000-0000-0000-0000-000000000000, Runtime=Wasm, Toolchain=Wasm, InvocationCount=1, IterationCount=1, IterationTime=250ms, MaxIterationCount=20, MemoryRandomization=True, MinIterationCount=15, RunStrategy=ColdStart, UnrollFactor=1, WarmupCount=0)
[2024/09/20 12:00:47][INFO] //    * The required .NET Core SDK version 9.0 for runtime moniker WasmNet90 is not installed.
[2024/09/20 12:00:47][INFO] 
```

**Comment by ilonatommy** [comment]
> noticed this one as well: https://helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-106599-merge-9f744b5760d349c987/x64.micro.net10.0.Partition0/1/console.a0f73dde.log?helixlogtype=result

I think this is understandable, considering that sdk 10 is expected to be used.

**Comment by ilonatommy** [comment]
Don't we need another case here:
https://github.com/dotnet/performance/blob/f6d0cff8bfb8bc0c3eec23b4b07c837e4c5f8abb/scripts/dotnet.py#L81?

We're using this method to set `framework` that we then use to set the env variable that is logged as net9:
https://github.com/dotnet/performance/blob/f6d0cff8bfb8bc0c3eec23b4b07c837e4c5f8abb/scripts/ci_setup.py#L511
This it not the full fix for sure, though

File: scripts/dotnet.py [hunk 2]
```diff
@@ -92,6 +92,8 @@ def get_target_framework_moniker(framework: str) -> str:
             return 'net8.0'
         if framework == 'nativeaot9.0':
             return 'net9.0'
+        if framework == 'nativeaot10.0':
+            return 'net10.0'
```
Comment (am11) [inline scripts/dotnet.py:96]: This case? I think we can hijack nativeaot9.0 for now to return net10.0 as well?

File: scripts/dotnet.py [hunk 3]
```diff
@@ -92,6 +92,8 @@ def get_target_framework_moniker(framework: str) -> str:
             return 'net8.0'
         if framework == 'nativeaot9.0':
             return 'net9.0'
```
Comment (am11) [inline scripts/dotnet.py:94]: Can try:
```suggestion
            return 'net10.0'
```

File: scripts/dotnet.py (hunk 2 above)
Comment (ilonatommy) [reply scripts/dotnet.py:96]: Let's try to add channel to [performance/azure-pipelines.yml](https://github.com/dotnet/performance/blob/f6d0cff8bfb8bc0c3eec23b4b07c837e4c5f8abb/azure-pipelines.yml#L149) - probably we should keep net9 if net8 is still there..

File: scripts/dotnet.py [hunk 4]
```diff
@@ -91,7 +91,9 @@ def get_target_framework_moniker(framework: str) -> str:
         if framework == 'nativeaot8.0':
             return 'net8.0'
         if framework == 'nativeaot9.0':
-            return 'net9.0'
+            return 'net10.0'
```
Comment (am11) [inline scripts/dotnet.py:94]:
```suggestion
            return 'net9.0'
```

File: scripts/dotnet.py (hunk 2 above)
Comment (am11) [reply scripts/dotnet.py:96]: Reverted, and update the pipeline.

**Comment by ilonatommy** [comment]
I re-run before the final touches were done, so in case it won't pass, we'll make another round.

**Comment by ilonatommy** [comment]
Waiting for the 2nd run to finish:
I'm not sure if we are even using it in this CI run but this place looks like a candidate for changes as well:
https://github.com/dotnet/performance/blob/f6d0cff8bfb8bc0c3eec23b4b07c837e4c5f8abb/src/tools/ResultsComparer/Data.cs#L131C31-L131C41

**Comment by am11** [comment]
BenchmarkDotNet apparently also has a lot of hardcoded monikers https://github.com/search?q=repo%3Adotnet%2FBenchmarkDotNet%20net9&type=code

Maybe we need to update BDN as well?

**Comment by am11** [comment]
Maybe this will help? At least for `PERFLAB_Framework` 😅

File: scripts/dotnet.py [hunk 5]
```diff
@@ -528,7 +530,7 @@ def run(self,
 FrameworkVersion = NamedTuple('FrameworkVersion', major=int, minor=int)
 @tracer.start_as_current_span("dotnet_get_framework_version") # type: ignore
 def get_framework_version(framework: str) -> FrameworkVersion:
-    groups = search(r".*(\d)\.(\d)$", framework)
+    groups = search(r".*(\d+)\.(\d+)$", framework)
```
Comment (am11) [inline scripts/dotnet.py:533]:
```suggestion
    groups = search(r".*?(\d+)\.(\d+)$", framework)
```

**Comment by am11** [comment]
Looks like regex required non-greedy when there is more than one digit in first match. Some sort of python gotcha, but this works (tested on godbolt)

**Comment by am11** [comment]
Next up:

> [2024/09/20 16:59:55][INFO] /datadisks/disk1/work/A2B508ED/p/dotnet/sdk/10.0.100-alpha.1.24468.12/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.TargetFrameworkInference.targets(166,5): error NETSDK1045: The current .NET SDK does not support targeting .NET 10.0.  Either target .NET 9.0 or lower, or use a version of the .NET SDK that supports .NET 10.0. Download the .NET SDK from https://aka.ms/dotnet/download [/datadisks/disk1/work/A2B508ED/w/B0B10936/e/performance/src/benchmarks/micro/MicroBenchmarks.csproj::TargetFramework=net10.0]

so it is correctly passing 10.0, but 10.0 SDK is only supporting .NET 9.. see https://github.com/dotnet/runtime/pull/106599#issuecomment-2364169080

**Comment by caaavik-msft** [comment]
We might need to get the BDN change to support targeting .NET 10 out and in this as well, similar to when we made this change last year:

- https://github.com/dotnet/BenchmarkDotNet/pull/2447
- https://github.com/dotnet/performance/pull/3428

File: scripts/channel_map.py [hunk 6]
```diff
@@ -3,8 +3,8 @@
 class ChannelMap():
     channel_map = {
         'main': {
-            'tfm': 'net9.0',
-            'branch': '9.0',
+            'tfm': 'net10.0',
+            'branch': '10.0',
```
Comment (caaavik-msft) [inline scripts/channel_map.py:7]: We should do this in a separate change as this shouldn't be a dependency to get https://github.com/dotnet/runtime/pull/106599 working. This also means we can spend some time to get the BDN changes in as well to support wasm on .NET 10.

Comment (caaavik-msft) [reply scripts/channel_map.py:7]: @am11 Can you remove the change to channel_map.py then we can merge this in?

File: azure-pipelines.yml [hunk 7]
```diff
@@ -146,6 +146,7 @@ jobs:
         runCategories: 'runtime libraries' 
         channels:
           - main
+          - nativeaot10.0
```
Comment (caaavik-msft) [inline azure-pipelines.yml:149]: Also if we can remove this line, we will add it once we are ready to support it

Comment (am11) [reply azure-pipelines.yml:149]:
```suggestion
```

File: scripts/channel_map.py (hunk 6 above)
Comment (am11) [inline scripts/channel_map.py:7]:
```suggestion
            'tfm': 'net9.0',
            'branch': '9.0',
```

**Review by caaavik-msft — APPROVED** [review]
LGTM

**Comment by am11** [comment]
@caaavik-msft thanks. Just curious, are those unsupported ones like (netcoreapp3) and desktop CLR ones (net4XX) channels still relevant / working? If not, maybe we should clean that up and keep it to lowest supported LTS.

**Comment by caaavik-msft** [comment]
Perhaps a good idea, I suppose there isn't much need for the older target frameworks. @DrewScoggins not sure if you had any other thoughts on if that might be something worth cleaning up.

**Comment by am11** [comment]
Judging by https://github.com/dotnet/performance/blob/d384536c59a6cda6b1a1524ff4c102cc4bfc5d8a/src/benchmarks/gc/GC.Infrastructure/GC.Infrastructure.Core/Configurations/Common.cs#L52 if we are actively disallowing COMPlus_ in favor of DOTNET_ environment vars, and those vars are essential for perf scenarios execution, then it means we are only supporting .NET 6+ (since DOTNET_ config functionality was added back then and COMPlus_ was the only supported prefix in previous versions).

**Comment by DrewScoggins** [comment]
Yeah, I was thinking about this. I am trying to come up with a realistic scenario that would involve us needing to go back and do performance testing on anything that old, and can't really get there. If the need did somehow arise, we can always just add it back in.

I think removing the obsolete TFMs seems like a good idea.

**Comment by DrewScoggins** [comment]
> Judging by
> 
> https://github.com/dotnet/performance/blob/d384536c59a6cda6b1a1524ff4c102cc4bfc5d8a/src/benchmarks/gc/GC.Infrastructure/GC.Infrastructure.Core/Configurations/Common.cs#L52
> 
> if we are actively disallowing COMPlus_ in favor of DOTNET_ environment vars, and those vars are essential for perf scenarios execution, then it means we are only supporting .NET 6+ (since DOTNET_ config functionality was added back then and COMPlus_ was the only supported prefix in previous versions).

That is for the GC perf tests and infrastructure, which is different than the existing microbenchmarks. However, I get your point.

**Pull request merged**